internal/util/parse/parse_test.go

```diff
@@ -182,3 +184,87 @@ func Test_pkgURLFromGHURL(t *testing.T) {
 		})
 	}
 }
+
+func Test_parseURL(t *testing.T) {
+	type expected struct {
+		repo    string
+		dir     string
+		version string
+	}
+	tests := map[string]struct {
+		ghURL string
+		expected
+	}{
+		"simple": {
+			ghURL:    "https://my.git-server.com/repository.git/directory@main",
+			expected: expected{repo: "https://my.git-server.com/repository", dir: "/directory", version: "main"},
+		},
+		"no ref": {
+			ghURL:    "https://github.com/GoogleContainerTools/kpt-functions-catalog.git/examples/apply-replacements-simple",
+			expected: expected{repo: "https://github.com/GoogleContainerTools/kpt-functions-catalog", dir: "/examples/apply-replacements-simple", version: "master"},
+		},
+		".git appears in the middle": {
+			ghURL:    "https://my-site.com/root.gitops.git/foo@main",
+			expected: expected{repo: "https://my-site.com/root.gitops", dir: "/foo", version: "main"},
+		},
+	}
+	for name, test := range tests {
+		test := test // capture range variable
+		t.Run(name, func(t *testing.T) {
+			ctx := printer.WithContext(context.Background(), printer.New(nil, nil))
+			repo, dir, version, err := parseURL(ctx, test.ghURL)
+			assert.NoError(t, err)
+			assert.Equal(t, expected{repo: repo, dir: dir, version: version}, test.expected)
+		})
+	}
+}
+
+func Test_GitParseArgs(t *testing.T) {
+	tests := map[string]struct {
+		ghURL    string
+		skip     bool
+		expected Target
+	}{
+		"git@ url with .git suffix": {
+			ghURL: "git@github.com:GoogleContainerTools/kpt.git",
+			expected: Target{Git: v1.Git{
+				Repo:      "git@github.com:GoogleContainerTools/kpt",
+				Directory: "/",
+				Ref:       "main",
+			},
+				Destination: "kpt"},
+			// in CI we get the error "git@github.com: Permission denied (publickey)" because the ssh key is not set up.
```

Comment:
Why does URL parsing require SSH keys and why would it get permission denied?

Reply:
`GitParseArgs` doesn't just parse the URL, it parses the URL and makes a call to [NewGitUpstreamRepo](https://github.com/GoogleContainerTools/kpt/blob/0e62e256d68c5daec50dd28e523566b84a31d4f2/internal/gitutil/gitutil.go#L141), somewhere in which it fetches the upstream repo.

Reply:
Thank you!